src/twisted/internet/tcp.py

```diff
@@ -853,6 +858,246 @@ def getPeer(self):
+        """
+        try:
+            self._fileDescriptor = self._fileFactory()
+        except (IOError, OSError) as e:
+            if e.errno == EMFILE:
+                self._log.failure("Failed to open reservation file.")
+                self._fileDescriptor = None
+            else:
+                raise
+
+
+    @_machine.output()
+    def _openReservationFile(self):
+        """
+        Use the C{_fileFactory} to claim a file descriptor.
+        """
+        self._createFD()
+
+
+    @_machine.output()
+    def _reopenReservationFile(self, fileObject):
+        """
+        The same as L{_openReservationFile}, except that it accepts a
+        C{fileObject} argument.
+
+        @param fileObject: ignored.
+        """
+        self._createFD()
+
+
+    @_machine.output()
+    def _closeReservationFile(self):
+        """
+        Close the reservation file.
+        """
+        self._fileDescriptor.close()
+        self._fileDescriptor = None
+
+
+    @_machine.output()
+    def _closeFile(self, fileObject):
+        """
+        Close the provided file object.
+
+        @param fileObject: The file object to close.
+        """
+        fileObject.close()
+
+
+    @_machine.output()
+    def _returnTrue(self):
+        """
+        Return True.
+
+        @return: L{True}
+        """
+        return True
+
+
+    @_machine.output()
+    def _returnFalse(self):
+        """
+        Return False.
+
+        @return: L{False}
+        """
+        return False
+
+
+    @_machine.output()
+    def _replaced(self, fileObject):
+        """
+        Return True.
+
+        @param fileObject: ignored.
+
+        @return: L{True}
+        """
+        if self._fileDescriptor is None:
+            self._openFailedWithEMFILE()
+            return False
+        return True
+
+
+    @_machine.output()
+    def _notReplaced(self, fileObject):
+        """
+        Return False.
+
+        @param fileObject: ignored.
+
+        @return: L{False}
+        """
+        return False
+
+
+    _unreserved.upon(
+        acquire, enter=_reserved, outputs=[_openReservationFile],
+        collector=_silenceOutputs)
+    _unreserved.upon(
+        replace, enter=_reserved,
+        outputs=[_closeFile, _reopenReservationFile, _replaced],
+        collector=_nthOutput(-1))
+    _unreserved.upon(
+        available, enter=_unreserved, outputs=[_returnFalse],
+        collector=_nthOutput(0))
+
+    _reserved.upon(
+        release, enter=_unreserved, outputs=[_closeReservationFile],
+        collector=_silenceOutputs)
+    _reserved.upon(
+        acquire, enter=_reserved, outputs=[],
+        collector=_silenceOutputs)
+    _reserved.upon(
+        replace, enter=_reserved,
+        outputs=[_notReplaced],
+        collector=_nthOutput(0))
+    _reserved.upon(
+        available, enter=_reserved, outputs=[_returnTrue],
+        collector=_nthOutput(0))
+    _reserved.upon(
+        _openFailedWithEMFILE, enter=_unreserved, outputs=[])
```

Comment:
This relies on Automat's transitioning to `_reserved` before the outputs for `_unreserved` --`replace`--> `_reserved` finish running.  That isn't a public API; see https://github.com/glyph/automat/issues/41.

Reply:
See also https://github.com/glyph/automat/issues/28